### ISSUE TITLE
How to configure apicurio studio with Keycloak and support authorization/access controls

### ISSUE THREAD

**Issue by nneul** [Apicurio/apicurio-studio#1820]
I don't want "everyone on the IDP" to have access to apicurio studio. All of the example setup all have apicurio-studio getting set up as a 'public' application with no authorization/access control capabilities. 

Can you provide an example configuration where it actually supports authorization.

**Comment by EricWittmann** [comment]
@carlesarnal let's discuss this soon.  It would be nice if this could be configured in Keycloak.  If not we'll need to add some basic RBAC support in Studio.

**Comment by nneul** [comment]
Note also, I'm not even looking for (at a minimum) any sort of differing levels of access beyond a simple 'on/off' access control. Certainly it would be nice to have user/admin access variants where 'admin' would grant ability to see all APIs regardless of sharing -- but that to me is a completely separate item.

When we tried turning on the authorization on keycloak, it just seemed to break everything, hence reason for report.

**Comment by carlesarnal** [comment]
Hi @nneul,

I have tested changing the client for the api to be a confidential client and configuring some resources using [Keycloak authorization services](https://www.keycloak.org/docs/latest/authorization_services/#_overview) and it works (api requests are rejected with the proper code). We probably lack some level of support in the UI for handling those codes). Could you please elaborate a bit more on what you did and how it broke the application?

Thanks

**Comment by nneul** [comment]
Let me give it another try and see what I get... we've done a bunch of different config changes since then trying to get a bunch of moving parts working together, so maybe behavior will be different now.

**Comment by nneul** [comment]
Everything immediately starts returning 'Forbidden'. 

Is there detail of what roles/secondary config is required - i.e. is studio expecting certain role names to be available? Also, should just 'confidential' be enabled, or should 'authorization' also be enabled?

I'm not seeing anything in the docs that really covers what is needed other than a generic "here, replace your entire keycloak realm with this import file" which isn't useful for an existing keycloak deployment.

**Comment by carlesarnal** [comment]
Yes, you need to enable authorization in Keycloak, as it's explained in the Keycloak [documentation](https://www.keycloak.org/docs/latest/authorization_services/#_getting_started_hello_world_enabling_authz_services) I've linked. Once the client has been made confidential and authorization has been enabled, you can configure authorization as desired. Using Keycloak Authorization Services you can configure the application at your convenience, from protecting a concrete endpoint to protecting a single design by using the API path etc. Everything is explained in the Keycloak documentation I've linked. As you can imagine, given how complex the domain is, even if we provide a default example realm, you're gonna likely need to customize the configuration.

**Comment by nneul** [comment]
The issue is - the default policy/etc. configuration for authorization in keycloak when you initially enable it on any app is effectively "grant for any users in this realm". I would fully expect that to result in 'everyone gets access to everything' -- however, the moment I turn on confidential+authorization on the app defined in KC, studio starts returning Forbidden message upon any attempted access. 

Clearly there is something missing, but I'm not sure what. I'm passing in the APICURIO_KC_CLIENT_SECRET and var along with the auth url realm and client name.

Is studio itself looking for some 'required' role/group/etc. that has to be set? Note that the 'forbidden' response is coming from studio url, unless it's being embedded in the keycloak response redirect.  I started from the docker example deployment, is there some other env var that needs to be included? I'll note that the docker example doesn't even include APICURIO_KC_CLIENT_SECRET in it's settings, so if there is something else required that could contribute here.

**Comment by carlesarnal** [comment]
First of all, let me make one thing clear that I should have explained earlier. The application is not fully ready to support Keycloak Authorization Services, there are some tweaks needed to the configuration and even dependencies that should be added to some poms (depending on how you're deploying/using the application).

Now direct to the point, in order to contact Kecloak, the client secret must be configured so the application can actually enforce the policies, this is very likely the reason why you're getting 403s from the API. The way it works is that the application will try to contact Keycloak to check the policies but, if there's no secret configured, it will fail with a 403, no matter the policies that you have in Keycloak (even if you don't have any of them).

Depending on how you're running the application, there are some options. In the Thorntail flavour, you need to modify the [keycloak.json](https://github.com/Apicurio/apicurio-studio/blob/master/platforms/thorntail/api/src/main/webapp/WEB-INF/keycloak.json) file with a configuration like [this](https://www.keycloak.org/docs/latest/authorization_services/index.html#_enforcer_filter) and also to add the secret, that will be used by the application when trying to enforce the policies.

If you're using Quarkus, the essence is exactly the same, but you'll need to configure everything in the `application.properties` file like [this](https://www.keycloak.org/docs/latest/authorization_services/index.html#_enforcer_filter).

As a recap, is **not** possible to do this right now with the existing images, you'll need to tweak the application a little bit and create your own images, at least until we get to the point of having this fully supported in Studio.

**Comment by nneul** [comment]
That's actually fairly helpful - let me dig in.

I'm currently using the vanilla docker images (apicurio/apicurio-studio-ui:latest-release, apicurio/apicurio-studio-api:latest-release, apicurio/apicurio-studio-ws:latest-release) launched explicitly (not using compose), but should be able to map in additional content/overrides - will see what I can figure out from above.

**Comment by nneul** [comment]
Related - one of our users is asking about assigning an 'admin' role of some sort - to let them be able to view ALL apis published in the system, and not just those that have been shared. Would the above steps be a prerequisite to that, or is that functionality even possible with the docker images?
